File: leet-code/206.反转链表.js
/*
 * @lc app=leetcode.cn id=206 lang=javascript
 *
 * [206] 反转链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */

var reverseList = function (head) {
  if (!head) return head

  let pre = null;
  let next = head;
  while (next) {
    const temp = next.next
    next.next = pre
    pre = next
    next = temp
  }
  return pre
};
// @lc code=end


// @after-stub-for-debug-begin
module.exports = reverseList;
// @after-stub-for-debug-end

File: leet-code-rs/src/solutions/206_反转链表.rs
/*
 * @lc app=leetcode.cn id=206 lang=rust
 *
 * [206] 反转链表
 */
#[allow(unused)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}
// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
#[allow(unused)]
impl Solution {
    pub fn reverse_list(head: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut pre: Option<Box<ListNode>> = None;
        let mut next = head;
        while let Some(mut node) = next.take() {
            let nxt = node.next.take();
            node.next = pre;
            pre = Some(node);
            next = nxt;
        }

        return pre;
    }
}
// @lc code=end

#[allow(unused)]
struct Solution;

#[cfg(test)]
mod test {
    #[allow(unused)]
    use super::*;

    #[test]
    fn test_reverse_list() {
        // assert_eq!(Solution::reverse_list(vec![]),[]);
        assert!(true)
    }
}
